keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,861 @@
+
+### Troubleshooting
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **How does this feature react if the API server and/or etcd is unavailable?**
+
+* **What are other known failure modes?**
+  For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Combine old and new fields
+
+Allow for 0 second Probe, with setting millisecond.
+
+Cons:
+Loosens validation.
+
+### v2 api for probe.
+
+I think this means a v2 API for Container therefore Pod.
+This seems too invasive.
+
+All Seconds fields become resource.Quantity instead of int32. This supports subdivision in a single field.
+
+### OffsetMilliseconds
+
+Use a negative offset, and combine with the existing field.
+
+Example:
+    Existing Field: int32 PeriodSeconds
+    New Field:      int32 PeriodOffsetMilliseconds
+
+    If I want to set to 0.5 seconds, 500 milliseconds,
+    PeriodSeconds <= 10 & PeriodOffsetmilliseconds <= -9500
```

Comment:
This is an example of a pair of values that could be ambiguous for round-tripping, forcing future APIs to store and track both pieces of data.

By contrast, if we had:
```yaml
....
   periodSeconds: 1
   earlyRecheckOffset:
      - unit: Millisecond 
        value: 500
...
```

and the v1 Pod API doesn't allow ambiguous values, then a future v2 Pod API can store a single field value behind the scenes
(eg: `periodMicroseconds = 500000u64`)

Reply:
Another option, still clamped to less than 1s:
By contrast, if we had:
```yaml
...
   periodSeconds: 0
   extraRecheckOffset:
      - unit: Millisecond 
        value: 500
...
```